File: Ult-4-Me v2/app/src/bridge/buttplug-server.js
/**
 * Buttplug v3 WebSocket server.
 * Accepts Buttplug-speaking clients and translates to Lovense HTTP API.
 */

const { WebSocketServer } = require('ws');
const { getActuators } = require('./device-db');

class ButtplugServer {
  constructor(lovense, { onLog = () => {} } = {}) {
    this.lovense = lovense;
    this._log = onLog;
    this.clientName = null;
    this._commandCount = 0;
    this._startTime = null;
    this._actuatorMap = {};  // `${devIdx},${actIdx}` -> { toyId, action }
    this._lastLevels = {};   // `${toyId},${action}` -> level
    this._wss = null;
  }

  _buildDeviceList() {
    const devices = [];
    this._actuatorMap = {};
    let devIdx = 0;

    for (const [tid, toy] of Object.entries(this.lovense.toys)) {
      const actuatorDefs = getActuators(toy.name, toy.functions);
      const scalarCmds = [];

      actuatorDefs.forEach(([lovenseAction, bpType], actIdx) => {
        scalarCmds.push({ FeatureDescriptor: lovenseAction, ActuatorType: bpType, StepCount: 20 });
        this._actuatorMap[`${devIdx},${actIdx}`] = { toyId: tid, action: lovenseAction };
      });

      devices.push({
        DeviceIndex: devIdx,
        DeviceName: `Lovense ${toy.name}`,
        DeviceMessages: { ScalarCmd: scalarCmds, StopDeviceCmd: {} },
        DeviceMessageTimingGap: 50,
      });
      devIdx++;
    }
    return devices;
  }

  async _handleMessage(type, body, id) {
    if (type === 'RequestServerInfo') {
      this.clientName = body.ClientName || 'Unknown';
      this._log(`Handshake: ${this.clientName}`);
      return [{ ServerInfo: { Id: id, ServerName: 'LvnsWatch Bridge', MessageVersion: 3, MaxPingTime: 0 } }];
    }
    if (type === 'RequestDeviceList') {
      const devices = this._buildDeviceList();
      this._log(`DeviceList: ${devices.length} device(s)`);
      return [{ DeviceList: { Id: id, Devices: devices } }];
    }
    if (type === 'StartScanning') {
      await this.lovense.refreshToys();
      const responses = [{ Ok: { Id: id } }];
      for (const dev of this._buildDeviceList()) {
        responses.push({ DeviceAdded: { Id: 0, DeviceIndex: dev.DeviceIndex, DeviceName: dev.DeviceName, DeviceMessages: dev.DeviceMessages } });
      }
      responses.push({ ScanningFinished: { Id: 0 } });
      const online = Object.values(this.lovense.toys).filter(t => t.status === 1).length;
      this._log(`Scan: ${online} online device(s)`);
      return responses;
    }
    if (type === 'StopScanning') return [{ Ok: { Id: id } }];

    if (type === 'ScalarCmd') {
      const devIdx = body.DeviceIndex || 0;
      for (const scalar of (body.Scalars || [])) {
        const actIdx = scalar.Index || 0;
        const level = Math.max(0, Math.min(20, Math.round((scalar.Scalar || 0) * 20)));
        const mapping = this._actuatorMap[`${devIdx},${actIdx}`];
        if (!mapping) continue;
        const levelKey = `${mapping.toyId},${mapping.action}`;
        if (this._lastLevels[levelKey] === level) continue;
        await this.lovense.sendAction(mapping.toyId, mapping.action, level);
        this._lastLevels[levelKey] = level;
        this._commandCount++;
        const toyName = this.lovense.toys[mapping.toyId]?.name || '?';
        this._log(`${toyName} ${mapping.action} ${level}/20`);
      }
      return [{ Ok: { Id: id } }];
    }
    if (type === 'VibrateCmd') {
      const devIdx = body.DeviceIndex || 0;
      for (const speed of (body.Speeds || [])) {
        const level = Math.max(0, Math.min(20, Math.round((speed.Speed || 0) * 20)));
        for (const [key, mapping] of Object.entries(this._actuatorMap)) {
          if (!key.startsWith(`${devIdx},`)) continue;
          const levelKey = `${mapping.toyId},${mapping.action}`;
          if (this._lastLevels[levelKey] === level) continue;
          await this.lovense.sendAction(mapping.toyId, mapping.action, level);
          this._lastLevels[levelKey] = level;
          this._commandCount++;
        }
      }
      return [{ Ok: { Id: id } }];
    }
    if (type === 'StopDeviceCmd' || type === 'StopAllDevices') {
      if (type === 'StopDeviceCmd') {
        const devIdx = body.DeviceIndex || 0;
        for (const [key, mapping] of Object.entries(this._actuatorMap)) {
          if (key.startsWith(`${devIdx},`)) {
            await this.lovense.stopToy(mapping.toyId);
            this._lastLevels[`${mapping.toyId},${mapping.action}`] = 0;
            break;
          }
        }
      } else {
        await this.lovense.stopAll();
        this._lastLevels = {};
      }
      this._commandCount++;
      this._log('Stop');
      return [{ Ok: { Id: id } }];
    }
    if (type === 'Ping') return [{ Ok: { Id: id } }];
    return [{ Ok: { Id: id } }];
  }

  start(port = 12345) {
    this._wss = new WebSocketServer({ host: '127.0.0.1', port });
    this._log(`Buttplug server on ws://127.0.0.1:${port}`);

    this._wss.on('connection', (ws) => {
      this.clientName = null;
      this._commandCount = 0;
      this._startTime = Date.now();
      this._lastLevels = {};
      this._log('Client connected');

      ws.on('message', async (raw) => {
        let messages;
        try { messages = JSON.parse(raw); } catch { return; }

        const responses = [];
        for (const msg of messages) {
          const [type, body] = Object.entries(msg)[0];
          const id = body.Id || 0;
          try {
            const resp = await this._handleMessage(type, body, id);
            if (resp) responses.push(...resp);
          } catch (e) {
            this._log(`Error: ${e.message}`);
            responses.push({ Ok: { Id: id } });
          }
        }
        if (responses.length > 0) {
          ws.send(JSON.stringify(responses));
        }
      });

      ws.on('close', () => {
        const elapsed = ((Date.now() - this._startTime) / 1000).toFixed(0);
        this._log(`Client '${this.clientName}' disconnected (${elapsed}s, ${this._commandCount} cmds)`);
        this.lovense.stopAll();
      });
    });
  }

  stop() {
    if (this._wss) {
      this._wss.close();
      this._wss = null;
    }
  }
}

module.exports = { ButtplugServer };

File: Prototype/bridge/buttplug_server.py
"""
Buttplug v3 WebSocket server.

Accepts clients (like the original Underwatch) and translates
Buttplug device commands to Lovense HTTP API calls.
Also allows our own app to call LovenseClient directly — the server
is for backward compat with any Buttplug-speaking app.
"""

import asyncio
import json
import time
import websockets
from websockets.asyncio.server import serve
from .device_db import get_actuators, KNOWN_DEVICES
from .lovense import LovenseClient


class ButtplugServer:
    def __init__(self, lovense: LovenseClient, on_log=None):
        self.lovense = lovense
        self._log = on_log or (lambda msg: None)
        self.client_name = None
        self._command_count = 0
        self._start_time = None
        # Maps (device_index, actuator_index) -> (toy_id, lovense_action)
        self._actuator_map = {}
        # Tracks last sent level per (toy_id, action) to skip duplicates
        self._last_levels = {}

    def _build_device_list(self):
        """Build Buttplug v3 device list and actuator mapping from connected toys."""
        devices = []
        self._actuator_map.clear()
        dev_idx = 0

        for tid, toy in self.lovense.toys.items():
            name = toy["name"]
            actuator_defs = get_actuators(name, toy.get("functions"))

            scalar_cmds = []
            for act_idx, (lovense_action, bp_type) in enumerate(actuator_defs):
                scalar_cmds.append({
                    "FeatureDescriptor": lovense_action,
                    "ActuatorType": bp_type,
                    "StepCount": 20,
                })
                self._actuator_map[(dev_idx, act_idx)] = (tid, lovense_action)

            devices.append({
                "DeviceIndex": dev_idx,
                "DeviceName": f"Lovense {name}",
                "DeviceMessages": {"ScalarCmd": scalar_cmds, "StopDeviceCmd": {}},
                "DeviceMessageTimingGap": 50,
            })
            dev_idx += 1

        return devices

    async def handle_client(self, ws):
        self.client_name = None
        self._command_count = 0
        self._start_time = time.time()
        self._last_levels.clear()
        self._log(f"Client connected from {ws.remote_address}")
        try:
            async for raw in ws:
                try:
                    messages = json.loads(raw)
                except json.JSONDecodeError:
                    continue
                responses = []
                for msg in messages:
                    try:
                        msg_type, body = next(iter(msg.items()))
                        msg_id = body.get("Id", 0)
                        resp = await self._handle_message(msg_type, body, msg_id)
                        if resp:
                            responses.extend(resp)
                    except Exception as e:
                        self._log(f"Error: {e}")
                        try:
                            responses.append({"Ok": {"Id": body.get("Id", 0)}})
                        except Exception:
                            pass
                if responses:
                    await ws.send(json.dumps(responses))
        except websockets.exceptions.ConnectionClosed:
            pass
        except Exception as e:
            self._log(f"Error: {e}")
        finally:
            elapsed = time.time() - self._start_time if self._start_time else 0
            self._log(f"Client '{self.client_name}' disconnected ({elapsed:.0f}s, {self._command_count} cmds)")
            try:
                await self.lovense.stop_all()
            except Exception:
                pass

    async def _handle_message(self, msg_type, body, msg_id):
        if msg_type == "RequestServerInfo":
            self.client_name = body.get("ClientName", "Unknown")
            self._log(f"Handshake: {self.client_name}")
            return [{"ServerInfo": {
                "Id": msg_id,
                "ServerName": "LvnsWatch Bridge",
                "MessageVersion": 3,
                "MaxPingTime": 0,
            }}]

        elif msg_type == "RequestDeviceList":
            devices = self._build_device_list()
            self._log(f"DeviceList: {len(devices)} device(s)")
            return [{"DeviceList": {"Id": msg_id, "Devices": devices}}]

        elif msg_type == "StartScanning":
            await self.lovense.refresh_toys()
            responses = [{"Ok": {"Id": msg_id}}]
            for dev in self._build_device_list():
                responses.append({"DeviceAdded": {
                    "Id": 0,
                    "DeviceIndex": dev["DeviceIndex"],
                    "DeviceName": dev["DeviceName"],
                    "DeviceMessages": dev["DeviceMessages"],
                }})
            responses.append({"ScanningFinished": {"Id": 0}})
            online = sum(1 for t in self.lovense.toys.values() if t["status"] == 1)
            self._log(f"Scan: {online} online device(s)")
            return responses

        elif msg_type == "StopScanning":
            return [{"Ok": {"Id": msg_id}}]

        elif msg_type == "ScalarCmd":
            dev_idx = body.get("DeviceIndex", 0)
            for scalar in body.get("Scalars", []):
                act_idx = scalar.get("Index", 0)
                value = scalar.get("Scalar", 0)
                level = max(0, min(20, round(value * 20)))

                mapping = self._actuator_map.get((dev_idx, act_idx))
                if not mapping:
                    continue

                toy_id, lovense_action = mapping
                level_key = (toy_id, lovense_action)
                if self._last_levels.get(level_key) == level:
                    continue

                await self.lovense.send_action(toy_id, lovense_action, level)
                self._last_levels[level_key] = level
                self._command_count += 1
                toy_name = self.lovense.toys.get(toy_id, {}).get("name", "?")
                self._log(f"{toy_name} {lovense_action} {level}/20")
            return [{"Ok": {"Id": msg_id}}]

        elif msg_type == "VibrateCmd":
            dev_idx = body.get("DeviceIndex", 0)
            for speed in body.get("Speeds", []):
                value = speed.get("Speed", 0)
                level = max(0, min(20, round(value * 20)))
                for (di, ai), (toy_id, lovense_action) in self._actuator_map.items():
                    if di != dev_idx:
                        continue
                    level_key = (toy_id, lovense_action)
                    if self._last_levels.get(level_key) == level:
                        continue
                    await self.lovense.send_action(toy_id, lovense_action, level)
                    self._last_levels[level_key] = level
                    self._command_count += 1
                    toy_name = self.lovense.toys.get(toy_id, {}).get("name", "?")
                    self._log(f"{toy_name} {lovense_action} {level}/20")
            return [{"Ok": {"Id": msg_id}}]

        elif msg_type in ("StopDeviceCmd", "StopAllDevices"):
            if msg_type == "StopDeviceCmd":
                dev_idx = body.get("DeviceIndex", 0)
                for (di, ai), (toy_id, lovense_action) in self._actuator_map.items():
                    if di == dev_idx:
                        await self.lovense.stop_toy(toy_id)
                        self._last_levels[(toy_id, lovense_action)] = 0
                        break
            else:
                await self.lovense.stop_all()
                self._last_levels.clear()
            self._command_count += 1
            self._log("Stop")
            return [{"Ok": {"Id": msg_id}}]

        elif msg_type == "Ping":
            return [{"Ok": {"Id": msg_id}}]

        else:
            return [{"Ok": {"Id": msg_id}}]

    async def start(self, host="127.0.0.1", port=12345):
        """Start the WebSocket server. Blocks until cancelled."""
        self._log(f"Buttplug server on ws://{host}:{port}")
        async with serve(self.handle_client, host, port):
            await asyncio.Future()
